include/execution/team_thread_md_range_iterator.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include "datatypes/impl/register_array.hpp"
+#include "macros.hpp"
+#include "policy.hpp"
+#include "void_iterator.hpp"
+#include <Kokkos_Core.hpp>
+#include <type_traits>
+
+namespace specfem {
+namespace execution {
+
+template <typename TeamMemberType, int... Ranks> class ThreadMDRangeIterator;
+
+template <typename TeamMemberType, int N1, int N2>
+class ThreadMDRangeIterator<TeamMemberType, N1, N2>
+    : public TeamThreadRangePolicy<TeamMemberType, int> {
+
+private:
+  using base_type = TeamThreadRangePolicy<TeamMemberType, int>;
+
+public:
+  constexpr static auto rank = 2; ///< Rank of the iterator (2 or 3)
+  using base_policy_type =
+      typename base_type::base_policy_type; ///< Base policy type. Evaluates to
+                                            ///< @c Kokkos::TeamThreadRange
+  using index_type = specfem::datatype::impl::RegisterArray<
+      int, Kokkos::extents<std::size_t, rank>,
+      Kokkos::layout_left>; ///< Underlying index type. This
+                            ///< index will be passed to the
+                            ///< closure when calling @ref
+                            ///< kokkos::parallel_for with
+                            ///< this iterator.
+
+  using execution_space =
+      typename base_type::execution_space; ///< Execution space type.
+
+  KOKKOS_INLINE_FUNCTION const index_type operator()(const int &i) const {
+    const int i1 = i / N2;
+    const int i2 = i % N2;
+    return index_type(i1, i2);
+  }
```

Comment:
The index ordering is dependent on where this this team lies. We need to have caching for default views on CPUs and coalescing on GPUs. 

Since `LayoutRight` is default for CPUs, to achieve caching on a 2D view I think we'd need to assign every column to individual thread i.e. i1 should increase first. 

For GPUs, LayoutLeft is default, to achieve coalescing on 2D view we'd need to assign every row to individual threads. So i2 needs should increase first. 
```suggestion
  KOKKOS_INLINE_FUNCTION const index_type operator()(const int &i) const {
   if constexpr (std::is_same_v<TeamMemberType::execution_space, Kokkos::DefaultHostExecutionSpace>) {
     const int i1 = i % N1;
     const int i2 = i / N1;
     return index_type(i1, i2);
   } else {
     const int i1 = i / N2;
     const int i2 = i % N2;
     return index_type(i1, i2);
   }
  }
```

Reply:
is `if constexpr (std::is_same_v<TeamMemberType::execution_space, Kokkos::DefaultHostExecutionSpace>)` preferred over `#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)` used in `chunked_domain_iterator.hpp`?

Reply:
Also, currently Views are not always using the default layout. For example, mesh quadrature View has hard-coded layout right.